### ISSUE TITLE
Use Symfony formtype as argument

### ISSUE THREAD

**Issue by nicodmf** [overblog/GraphQLBundle#453]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Version/Branch   | all

As nelmio api rest doc has, this bundle could use Symfony FormType as argument. The advantages : 

- less argument declaration
- automatic validation and error handling
- best integration in the framework

All basic FormType could be automaticaly defined as GraphQL types in the default configuration and a form introspection could generate GraphQL additionnals types.

**Comment by mcg-web** [comment]
hi @nicodmf, this could be a nice feature indeed, do you want to work on this?

**Comment by nicodmf** [comment]
Hi, i can, have you little informations on where to begin ?

**Comment by Vincz** [comment]
Hi! I was thinking about it, and I don't really see what is possible to do about this. The Symfony Form provide a way to display and handle HTML forms with the "view" part, the "normalization part", etc... It displays widgets and normalize the data, etc...
With GraphQL, we don't have forms. We have mutations. And GraphQL already describes the way this mutations should work. We don't have the "client" part (as it is probably build using a javascript framework).
You could still convert Form type to GraphQL Input, but I can't really see the benefit as you will have to express your form as "Widgets" before converting it to "GraphQL Type". It just seems to be an unnecessary additional step. 
At the end, I think the only thing you could work on, is to ease the validation process of the mutation inputs, but it's a just a matter of using the validation component with validation classes. 
You can also check this about mutation input & validations : https://github.com/overblog/GraphQLBundle/blob/annotations/docs/annotations/arguments-transformer.md

Hope it will help

**Comment by nicodmf** [comment]
I don't think about validation in the first time but more on args generation, like with the builder https://github.com/overblog/GraphQLBundle/blob/master/docs/definitions/builders/args.md. As the validation is included in form this could be surelly used in the second time.

Concerning the form component, even if it could be used as html generator, the base concept is biggest :

> The Form component is a tool to help you solve the problem of allowing end-users to interact with the data and modify the data in your application. And though traditionally this has been through HTML forms, the component focuses on processing data to and from your client and application, whether that data be from a normal form post or from an API.

IMO, graphql don't describe server treatment, just client and server inter communications with a simple, easy and comprehensible API, but nothing about consistant validation like data dependencies between 2 fields, or more obvious uniqness for a field. This rules should be treat by the server. As Symfony form can represents (the schema) and validate (the response) a complex structure of data, it seems to me interessant to use this power instead recode a complete definition and validation structure from zero. And for Developper experience, this will be too better, and the developper doesn't have to learn new concept for a same process.

Error handling is really another point which could be improve and surelly more consistantly with form integration :

- https://facebook.github.io/graphql/June2018/#sec-Errors
- https://blog.apollographql.com/full-stack-error-handling-with-graphql-apollo-5c12da407210#1cbd

Thanks for your link, this is usefull and demonstrate that a complex validation come from symfony way.

**Comment by nicodmf** [comment]
A POC which treat basic form is available in #461 

The usage is by exemple : 

```yaml
Query:
    type: object
    config:
        fields:
            GetSite:
                type: "String"
                resolve: "@=resolver('GetSite', [args])"
                form: App\Form\AFormType
                args:
                    id: { type: "String!", defaultValue: null }
```

This add args from form (for moment basic fields: text, number) and merge with args

The problems not resolve which demands yours advices :
- to add types discovered in the form to the schema (i don't find schema and manipulate it in the ConfigProcessor)
- a more efficiency method could be to change the generated classes (that i would to do in the beginning, but, if i understand the actual process, the first compilation doesn't access the containerInterface or services and a form inspection must have service access)

**Comment by akomm** [comment]
@nicodmf you are right about what symfony states about the component, though I agree with @Vincz that the Form component is in reality way to HTML-centric. Just check this out: https://github.com/symfony/form/blob/master/Form.php. 

At some point I was deciding whether I will use it in API context based on the quoted official statement, but its simply is not really that generic. I would not recommend use it in a different context than HTML.

**Comment by nicodmf** [comment]
submit, handleRequest are not only for html puposes. clickedButton yes, but why not used a good and many tested library when just a function doesn't usefull. 

It's the fourth projet that i lead  with this bundle, and really the error handling is to implements. The 3 first projets i used a dedicated graphql type to be comprehensible by our developpers (back and front). The Form usage is the last implementation that we used, but with down side, we should define same data in many place : 
- form to have a simple validation
- entity to save and load db 
- dto when data are not doctrine safe
- validators to check type
- graphql to provide schema

At the beginning of the last projet we used validator to validate data, but it is clearly a verbose way and the form component is more efficient to handle that.

This Idea is not just a idea from a developper which use this bundle for the first time.

IMO, if the big library as Nelmio api and FOS Rest (https://symfony.com/doc/master/bundles/FOSRestBundle/2-the-view-layer.html#forms-and-views) use it form request treatment this isn't just for html.

With form, we win : 
- a structure with automatics types
- an event system
- a direct link to doctrine
- an validation process
- a better integration in Symfony
- a better DX (developper experience)

This advantages are bigger, and permits to develop an API faster with a real control on the structure validated.

But, i understand maybe why this is not easy : as the schema is generated before access to services, this bundle doesn't not access to many Symfony advantages and rewrite this behavior could be a heavy challenge.

**Comment by akomm** [comment]
Big library, good library and a library that fits are distinct things to me. We should not argue how big something is, but how does it relate to this project and how does it fit into it. 

Vincz has made some good points as to why exactly it does not fit. I agree with his opinion. 

Also what you have posted here is simply done with resolver + builder and can be even a composer package if you want so. You would not have to configure more in the yaml, just other keywords (builder, builderConfig). So what exactly does in the way how currently this bundle work prevents you from doing it the resolver + builder way? You could this way use all the symfony advantages you want (resolver as service).

**Comment by nicodmf** [comment]
To much boilerplate i think.

**Comment by akomm** [comment]
> With form, we win :
>  1. a structure with automatics types
>  2. an event system
>  3. a direct link to doctrine
>  4. a validation process
>  5. a better integration in Symfony
>  6. a better DX (developper experience)
>  7. receive a validated data Object in the resolvers

1. You don`t simply take a structure for automatic types if it does not make sense, as Vincz explained why.
2. Linux kernel also has event system. How does linux kernel fit into this project?
3. Why should there be a direct link to doctrine?
4. We could have this with the validation component without all the html-centric widget- system of forms
5. How that? Just because you hard-code more from symfony/* does not make it better integrated, if it does not make sense to integrate it.
6. How that?
7. We can do this with validation component too. 

> Why ? Too much boilerplate, i think.

Why do you have to copy all the form code you PR'ed for every input argument?

**Comment by Vincz** [comment]
> I woulnd't replace mutations and queries but just prepare args, and auto respond if args are wrong, nothing more.

It's what I implemented in the [Arguments Transformer](https://github.com/overblog/GraphQLBundle/blob/annotations/src/Transformer/ArgumentsTransformer.php) in the `annotations` branch. It just prepares arguments (turn submitted data into PHP objects), validate them and respond with errors if it fails. From this point, as we are mapping the data to PHP objects, you can do whatever you want with it (the object can be an entity, you can add validation on it, etc...). 
I can't think about a situation where I would need an events system in the process as GraphQL Input are already formatted data structure, but if you have one...

I agree with @akomm on this, why add such a big layer of complexity, just to have validation on the input?

But if you have concrete examples from your projects, we could study this together.

**Comment by murtukov** [comment]
@nicodmf 
First of all I am glad to tell you, that we are currently working on the Validation feature, which will allow you to use validation constraints directly in the type definition (currently only yaml).

You could use it like this:
```yaml
Mutation:
    type: object
    config:
        fields:
            createUser:
                type: "String"
                resolve: "@=mut('createUser', [args, validator])"
                args:
                    groups:
                        type: "[String]"
                    userId:
                        type: Int!
                        validation:
                            - EntityExists: App\Entity\User
                    username:
                        type: String!
                        validation:
                            - Length:
                                min: 6
                                max: 32
                                minMessage: 'create_user.username.length.min'  # will be translated
                            - NotBlank:
                                message: 'Something is empty'
                    password:
                        type: String!
                        validation:
                            - Length: { min: 6, max: 32 }
                            - IdenticalTo:
                                  propertyPath: 'passwordRepeat'
                                  message: 'Passwords are not equal!'
                    passwordRepeat:
                        type: String!
```

This is only scratching the surface, as there will be such features as **linking classes**, to reuse the validation constraints declared on your entities, **cascade** validation, **custom constraints**, **groups**, **group sequeces**, **error formating** etc. With this, the need of manual validation and error handling automatically eliminates.

Secondly I don't think using the Form component is a good idea, as it was designed to work with HTML. I know, that the component can do more, than just working with HTML, but this is not it's main goal. GraphQL has nothing to do with views (e.g. HTML). It has it's own data description tool - **Schema** and **Types**. You can consider them as "forms for GraphQL".

Basically you are suggesting to use forms only to automatically declare arguments, but this already could  be done with argument builders.

Finally I am planning to work on Data Transformer feature in the near future. So I don's see any use of what you are suggesting.

**Comment by nicodmf** [comment]
@murtukov, @Vincz, i'm happy to see a complete integration of arguments and validation. It's effectivly better to integrate them separatly with there new components than use form.

For the event system, i'm sure we could find at least one exemple, as with form :

By example, if you send contact (mutation SendContact) with subject and optional subsubject with this rules :
- subsubject not exist if subject is *one*
- subsubject is required and other rules (length, string without email or in array of strings, etc...) if subject is *two*

This could be done with validation, if object validation are , but if you must search the list of subsubject in the database, this could be easiest with events.

If i take the murtukov example, the list of groups could be better with an event system which links each group to a database objects rather than to strings.

But this bundle evolve in the good way and the event system coulds be surelly implemented in other way.

**Comment by murtukov** [comment]
@nicodmf I didn't really get want you mean. Could you provide some examples maybe?

**Comment by drmax24** [comment]
> @nicodmf First of all I am glad to tell you, that we are currently working on the Validation feature, which will allow you to use validation constraints directly in the type definition (currently only yaml).
> 
> You could use it like this:
> 
> ```yaml
> Mutation:
>     type: object
>     config:
>         fields:
>             createUser:
>                 type: "String"
>                 resolve: "@=mut('createUser', [args, validator])"
>                 args:
>                     groups:
>                         type: "[String]"
>                     userId:
>                         type: Int!
>                         validation:
>                             - EntityExists: App\Entity\User
>                     username:
>                         type: String!
>                         validation:
>                             - Length:
>                                 min: 6
>                                 max: 32
>                                 minMessage: 'create_user.username.length.min'  # will be translated
>                             - NotBlank:
>                                 message: 'Something is empty'
>                     password:
>                         type: String!
>                         validation:
>                             - Length: { min: 6, max: 32 }
>                             - IdenticalTo:
>                                   propertyPath: 'passwordRepeat'
>                                   message: 'Passwords are not equal!'
>                     passwordRepeat:
>                         type: String!
> ```
> 
> This is only scratching the surface, as there will be such features as **linking classes**, to reuse the validation constraints declared on your entities, **cascade** validation, **custom constraints**, **groups**, **group sequeces**, **error formating** etc. With this, the need of manual validation and error handling automatically eliminates.
> 
> Secondly I don't think using the Form component is a good idea, as it was designed to work with HTML. I know, that the component can do more, than just working with HTML, but this is not it's main goal. GraphQL has nothing to do with views (e.g. HTML). It has it's own data description tool - **Schema** and **Types**. You can consider them as "forms for GraphQL".
> 
> Basically you are suggesting to use forms only to automatically declare arguments, but this already could be done with argument builders.
> 
> Finally I am planning to work on Data Transformer feature in the near future. So I don's see any use of what you are suggesting.

I have an error:
Constraint class 'Symfony\Component\Validator\Constraints\EntityExists' doesn't exist.

When trying to do smth you proposed:
>                     userId:
>                         type: Int!
>                         validation:
>                             - EntityExists: App\Entity\User

Is it possible to check if entity exists via validation?
